services/audit/display.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2024 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package audit
+
+import (
+	"context"
+	"fmt"
+
+	asymkey_model "code.gitea.io/gitea/models/asymkey"
+	audit_model "code.gitea.io/gitea/models/audit"
+	auth_model "code.gitea.io/gitea/models/auth"
+	"code.gitea.io/gitea/models/db"
+	git_model "code.gitea.io/gitea/models/git"
+	organization_model "code.gitea.io/gitea/models/organization"
+	repository_model "code.gitea.io/gitea/models/repo"
+	secret_model "code.gitea.io/gitea/models/secret"
+	user_model "code.gitea.io/gitea/models/user"
+	webhook_model "code.gitea.io/gitea/models/webhook"
+)
+
+type cache = map[audit_model.ObjectType]map[int64]TypeDescriptor
+
+func FindEvents(ctx context.Context, opts *audit_model.EventSearchOptions) ([]*Event, int64, error) {
+	events, total, err := audit_model.FindEvents(ctx, opts)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return fromDatabaseEvents(ctx, events), total, nil
+}
+
+func fromDatabaseEvents(ctx context.Context, evs []*audit_model.Event) []*Event {
+	c := cache{}
+
+	users := make(map[int64]TypeDescriptor)
+	for _, systemUser := range []*user_model.User{
+		user_model.NewGhostUser(),
+		user_model.NewActionsUser(),
+		user_model.NewCLIUser(),
+		user_model.NewAuthenticationSourceUser(),
+	} {
+		users[systemUser.ID] = typeToDescription(systemUser)
+	}
+	c[audit_model.TypeUser] = users
+
+	events := make([]*Event, 0, len(evs))
+	for _, e := range evs {
+		events = append(events, fromDatabaseEvent(ctx, e, c))
+	}
+	return events
+}
+
+func fromDatabaseEvent(ctx context.Context, e *audit_model.Event, c cache) *Event {
+	return &Event{
+		Action:    e.Action,
+		Actor:     resolveType(ctx, audit_model.TypeUser, e.ActorID, c),
+		Scope:     resolveType(ctx, e.ScopeType, e.ScopeID, c),
+		Target:    resolveType(ctx, e.TargetType, e.TargetID, c),
+		Message:   e.Message,
+		Time:      e.TimestampUnix.AsTime(),
+		IPAddress: e.IPAddress,
+	}
+}
+
+func resolveType(ctx context.Context, t audit_model.ObjectType, id int64, c cache) TypeDescriptor {
+	oc, has := c[t]
+	if !has {
+		oc = make(map[int64]TypeDescriptor)
+		c[t] = oc
+	}
+
+	td, has := oc[id]
+	if has {
+		return td
+	}
+
+	switch t {
+	case audit_model.TypeSystem:
```

Comment:
We should not store the ID or at least we need other identify like username, reponame and etc. We should not load the object from the database with the recorded id because it may be deleted.

Reply:
See line 119 and 129